feat(experiments): enable results breakdown for the new engine.

rodrigoi opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

The `ResultsQuery` and `ExploreButton` (used on `ResultsTab` and `ChartModal`) are incompatible with the new engine query and response format, so funnel and trends metric breakdowns are enabled only for legacy experiments.

### Legacy Implementation

The legacy experiment metrics queries (`ExperimentFunnelsQuery` and `ExperimentTrendsQuery`) work by performing the funnel or trend query, processing that information, and returning significance data.
In the process, the insight response is cleaned up, removing breakdowns that don't match the experiment's feature flag variations. This cleaned-up response is attached to the response payload as `insight`, which enables the breakdown chart.

### Query Component

The breakdown view is powered by the `Query` component, which receives the funnel or trend query, and a context object with additional configuration as `context`. In this `context`, we pass the response `insight` as a cached insight result, so the query never runs from the client.

| _Baseline_ added automatically | loading results from external query |
| - | - |
| <img width="1419" alt="image" src="https://github.com/user-attachments/assets/2feb8602-b74b-4b0d-8a3b-3e51f2e60da9" /> | <img width="1418" alt="image" src="https://github.com/user-attachments/assets/a536ce47-9f4b-464b-b3a5-7a0a085d6bac" /> |


<!-- Who are we building for, what are their needs, why is this important? -->

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

## Changes

### ExperimentQuery

For the new engines, the `ExperimentQueryResponse` still supports the `insight` but always returns an empty array. Enabling this field would provide the necessary context to the `Query` component to treat the query as cached. Still, it will limit our ability to introduce changes to the experiment's backend.

### Querying from the client

Instead of breaking the contract with `ExperimentQueryResponse` and `ExperimentMetricResponse`, we'll create a query on the client with the correct exposure criteria. Passing the query directly to the `Query` component would provide the best UX (for example, having a progress indicator), but it would make it difficult to limit the results to the experiment variants.

So, we are adding a new logic to handle creating and performing the query, filtering the results, and presenting them to the `Results` component in a format that satisfies the cached response. We'll lose some UX benefits.

| Single Metric | Chart Dialog |
| - | - |
| <img width="1466" alt="image" src="https://github.com/user-attachments/assets/212ee081-3a69-4b05-9415-610c739196b6" /> | <img width="1719" alt="image" src="https://github.com/user-attachments/assets/bab1bae2-c030-4720-a286-47d5e43c5ec7" /> |

### Splitting Code Paths

Following the pattern of other PRs, we separate _legacy_ code from the new implementation. We are also marking _legacy_ components as `@deprecated` so we can have a visual indication of their condition.

### Composing the UI

To support the single metric view and the `ChartModal`, we are using a _render prop_ pattern (alternatives include shared context or `cloneElement`) so we can render the components with the necessary structural differences, while providing the `props` they need. We are focusing on having pure components and a top-level logic mount.

### The Sneaky Refactor

This PR includes a new module with composable helper functions that we can use to transform and augment metrics. The implementation in the `utils` module is not extensible to support all the use cases we have (especially for funnels with an exposure criterion). We'll refactor the rest of the code in future PRs.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've added or updated the docs
- [ ] I've reached out for help from the docs team
- [x] No docs needed for this change

## How did you test this code?

The metric utilities have unit tests:

```bash
pnpm --filter=@posthog/frontend jest src/scenes/experiments/metricQueryUtils.test.ts
```

the rest...

![cat-type-small](https://github.com/user-attachments/assets/5db22bcb-3e92-432a-a52a-19fd404093f3)

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

rodrigoi committed:
feat: enable single metric view for experiment query.

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

rodrigoi committed:
fix(experiments): creating a new logic for the results query.

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

rodrigoi committed:
feat(experiments): breakdown logic updates to support custom exposure criterias.

rodrigoi committed:
feat(experiments): setting up trends metrics.

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

rodrigoi committed:
fix(experiments): some updates to the breakdown logic.

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rodrigoi committed:
feat(experiments): adding breakdown to details modal.

rodrigoi commented on frontend/src/scenes/experiments/experimentResultBreakdownLogic.ts:
```diff
@@ -0,0 +1,216 @@
+        metric: null as ExperimentMetric | null,
+    } as ExperimentResultBreakdownLogicProps),
+
+    key(
+        ({ experiment, metric }: ExperimentResultBreakdownLogicProps) =>
+            `${experiment.id}-${metric.kind}-${metric.name}`
+    ),
```
these will never be null.

rodrigoi commented on frontend/src/scenes/experiments/MetricsView/ChartModal.tsx:
```diff
@@ -20,6 +20,7 @@ interface ChartModalProps {
     isSecondary: boolean
     result: any
```
out of scope.

rodrigoi commented on frontend/src/scenes/experiments/ExperimentView/components.tsx:
```diff
@@ -134,7 +137,63 @@ export function ResultsTag({ metricIndex = 0 }: { metricIndex?: number }): JSX.E
+
+    const { query, results } = breakdownResults
+
+    const fakeInsightId = Math.random().toString(36).substring(2, 15)
```
keeping this, will probably refactor later.

rodrigoi committed:
fix(experiments): applying automatic feedback.

rodrigoi commented on frontend/src/scenes/experiments/experimentResultBreakdownLogic.ts:
```diff
@@ -0,0 +1,216 @@
+export type BreakDownResults = {
+    query: FunnelsQuery | TrendsQuery
+    results: FunnelStepsResults
+}
```
updated in 6a68505623ba264da619ef929b6747809d85c2a2

rodrigoi commented on frontend/src/scenes/experiments/experimentResultBreakdownLogic.ts:
```diff
@@ -0,0 +1,216 @@
+                        let results = response.results as FunnelStepsResults
+
+                        /**
+                         * we need to filter the results to remove any non-variant brakedowns
```
updated in 6a68505623ba264da619ef929b6747809d85c2a2

rodrigoi commented on frontend/src/scenes/experiments/MetricsView/ChartModal.tsx:
```diff
@@ -45,12 +47,9 @@ export function ChartModal({
-                    <ExploreButton result={result} />
-                </div>
-            )}
+            <div className="flex justify-end">
+                <ExploreButton result={result} />
+            </div>
```
updated in 6a68505623ba264da619ef929b6747809d85c2a2

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rodrigoi committed:
fix(experiments): temporarily hiding the explore button for non-legacy metrics.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

rodrigoi committed:
refactor: great unification of transformation functions.

rodrigoi committed:
refactor: some functional composition to create a query.

jurajmajerik commented:
The Funnel insight looks great! 🌪️

I'm not so sure about the Trend insight though. It won’t be accurate when Winsorization is used, since it doesn’t filter out extreme values. I’d love for @andehen to weigh in here too, he has ideas on what might be useful to show, and a trend chart might not be the best fit.

To keep things moving, I think it makes sense to limit this PR to the Funnel insight only.

rodrigoi committed:
fix: correcting some typos.

rodrigoi committed:
test: adding tests for the metricToFilter refactor.

rodrigoi commented on frontend/src/scenes/experiments/metricQueryUtils.ts:
```diff
@@ -0,0 +1,352 @@
+}
+
+export const addExposureToMetric =
+    (expposureEvent: EventsNode) =>
```
This happens because I don't vibe code... bacaedfdb1c941f80ef0783c40762b9b146139df

rodrigoi committed:
tests(experiments): getQuery and getFilter pass the metricToQuery and metricToFilter test suite.

rodrigoi committed:
feat: adding outlier handling to query transformation functions.

andehen reviewed (commented):
Nice! Tested locally and works well here 👍 

> There's a lot of duplicated code to transform metrics and results into queries or viz nodes. This needs a bit of a cleanup before we can completely fix the explore button.

Great! Great to consolidate and clean up this logic. 

Some thoughts:
- We should probably add a tool tip to documentation where we explain why funnel breakdown numbers might differ from experiment numbers (mostly due to the $mutliple handling, f.ex). We are going to get these questions.
- I wonder if we should not display this for "mean" metrics as of now. We will never get the results to match with trends queries, and no-one has requested the feature for mean metrics yet. I think it's much less useful for those metrics where as funnel breakdowns does provide additional insight. It does make sense to add a delta chart timeseries breakdown, but that will require a completely different implementation (we have to do it ourselves then, can't use product analytics)
- It also seems to not work currently for mean metrics in my case I just got this:
![Screenshot 2025-06-10 at 15 44 21](https://github.com/user-attachments/assets/dd425846-5308-43ab-96b3-bcd34c2508d3)
I guess it's due to the feature flag information not being present on the events themselves. But back to my point earlier, I think we should skip support for mean metrics here.

rodrigoi committed:
fix: rolling back exception for events node check.

rodrigoi committed:
refactor: moving files arround to make it more manageable.

rodrigoi committed:
refactor: simplifyng result breakdown logic.

rodrigoi committed:
feat: fixing the explore as insight button.

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

rodrigoi committed:
fix: binding to logic in the top level component.

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rodrigoi committed:
feat: composing the ChartModal.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rodrigoi committed:
refactor: taking composition one step further.

rodrigoi commented on frontend/src/scenes/experiments/ExperimentView/ExperimentView.tsx:
```diff
@@ -62,30 +71,52 @@ const ResultsTab = (): JSX.Element => {
                                     </>
+                                ) : (
+                                    /**
+                                     * altough we don't have a great typeguard here, we know that the result is a CachedExperimentQueryResponse
+                                     * because we're only showing results for experiment queries (legacy check)
```
out of scope

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rodrigoi commented on frontend/src/scenes/experiments/components/ResultsBreakdown/resultsBreakdownLogic.ts:
```diff
@@ -0,0 +1,199 @@
+                             * we filter from the series all the breakdowns that do not map
+                             * to a feature flag variant
+                             */
+                            results = (results as TrendResult[]).filter(
+                                (series) =>
+                                    series.breakdown_value !== undefined &&
+                                    variants.includes(series.breakdown_value as string)
+                            )
```
we'll fix on a different PR.

rodrigoi commented:
@andehen, this is ready for another review! I've updated the code to support only funnel metrics, rebased to include the frequentist changes, and refactored the components so we can have a composable interface. I'll get loading states and information tooltips in another PR; this has gotten quite large.

andehen commented on frontend/src/scenes/experiments/ExperimentView/ExperimentView.tsx:
```diff
@@ -41,10 +44,16 @@ const ResultsTab = (): JSX.Element => {
         statsMethod,
     } = useValues(experimentLogic)
+    /**
+     * we still use the legacy metric results here. Frequentist results are loaded
```
nit: This will also be the case bayesian soon
```suggestion
     * we still use the legacy metric results here. Results on the new format are loaded
```

andehen reviewed (commented):
Looks good! Great consolidation of logic in `metricQueryUtils` 👍 

One thing i noticed is that the details table is removed for mean metrics. It just displays an empty box now:
<img width="1282" alt="Screenshot 2025-06-12 at 11 21 41" src="https://github.com/user-attachments/assets/91c2ee5d-8d3c-4df2-a68d-7ef68cede4d3" />

We should still render that as we do currrently
<img width="1227" alt="Screenshot 2025-06-12 at 11 33 51" src="https://github.com/user-attachments/assets/210bd375-822c-4a10-bb37-0b9b1f53f868" />

rodrigoi committed:
fix: updating comments about the new result format.

rodrigoi committed:
fix: allow child components to narrow the interface and force type safety.

rodrigoi commented on frontend/src/scenes/experiments/components/ResultsBreakdown/resultsBreakdownLogic.ts:
```diff
@@ -0,0 +1,195 @@
+}
+
+/**
+ * This logic only works with modern engines, like bayesian and frequentist.
```
updated here: 6daa168f0fd869adfb40552b2288b359706cbfbb

rodrigoi commented:
@andehen ! Thanks for the review. I think that broke on the last rebase that had some conflicts. I've updated both `ExperimentView` and `CharModal` to fix this, and included a little extra fix that I was planning for a future PR to improve the type safety of the children components, but it made sense to include here, so the changes you make more sense.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

andehen reviewed (commented):
Nice job on this! A big one to get over the line.

Works perfectly on my test case with the default exposure criteria! 👍 

I did ran into an issue when using a custom pageview one. The number of exposures matches, but for some reason the conversion calculation is different:

<img width="1217" alt="Screenshot 2025-06-12 at 18 19 24" src="https://github.com/user-attachments/assets/6444f136-efcf-47e4-b91c-7da60e667fb9" />

Didn't have time to look closer now. But something we should figure out I think.

I also noticed that the breakdown for the query with default exposure criteria uses "Feature: <key>". We should use the "Feature flag response" value here to be consistent with the query runner. For events other than `$feature_flag_called`, we still need to use "Feature: <key>".

And also, I think we should change the attribution type on the funnel to "First step". To be consistent with the query runner.

andehen commented on frontend/src/scenes/experiments/metricQueryUtils.ts:
```diff
@@ -0,0 +1,369 @@
+                    dateRange,
+                    interval: trendsInterval,
+                    trendsFilter,
+                    ...getOutlierHandling(meanMetric),
```
I don't think "outlier handling" is relevant for Trend queries?

rodrigoi committed:
feat: hiding the results breakdown behind a feature flag.

rodrigoi committed:
Merge branch 'experiments/32661-experiment-query-breakdown-vizualization' of github.com:PostHog/posthog into experiments/32661-experiment-query-breakdown-vizualization
